src/tests/integration/monitor-tabs.test.ts

```diff
@@ -0,0 +1,200 @@
+    it("Attaches a tab to another window and updates the store's state", () => {
+      const [callbackOnAttached] = mocks.chrome.tabs.onAttached.addListener.firstCall.args
+
+      const tabId = 15
+      const attachInfo = {
+        id: 15,
+        windowId: 2,
+        newWindowId: 3,
+        active: false,
+        url: 'https://slam.com/abc',
+      }
+
+      callbackOnAttached(tabId, attachInfo)
+
+      const state = mocks.getState()
+      const tab = state.tabs.get(tabId)!
+
+      expect(tab.id).to.equal(attachInfo.id)
+      expect(tab.windowId).to.equal(attachInfo.newWindowId)
+      expect(tab.domain).to.equal('slam.com')
+      expect(tab.url).to.equal(attachInfo.url)
+    })
+  })
+
+  describe('chrome.tabs.onActivated', () => {
+    it('sets the specified tab as active for a given window and sets any other tabs of that window to be inactive', () => {
+      const [callback] = mocks.chrome.tabs.onActivated.addListener.firstCall.args
+      const activeInfo: chrome.tabs.TabActiveInfo = { tabId: 13, windowId: 1 }
+      callback(activeInfo)
+
+      const { tabs } = mocks.getState()
+      expect(tabs.get(13)).to.have.property('active', true)
+      expect(tabs.get(12)).to.have.property('active', false)
+    })
+
+    it('has no effect for active tabs of other windows', () => {
+      const { tabs } = mocks.getState()
+      expect(tabs.get(17)).to.have.property('active', true)
+    })
+  })
+
+  describe('chrome.tabs.onReplaced', () => {
+    it("replaces a tab (create and remove) and updates store's state", () => {
+      expect(mocks.chrome.tabs.onCreated.addListener).to.have.callCount(1)
+      const [callback] = mocks.chrome.tabs.onCreated.addListener.firstCall.args
+
+      callback({
+        id: 11,
+        windowId: 1,
+        active: false,
+        url: 'https://foo.com/abc',
+      })
+
+      const state = mocks.getState()
+      const tab = state.tabs.get(11)!
+
+      expect(tab.id).to.equal(11)
+      expect(tab.windowId).to.equal(1)
+      expect(tab.domain).to.equal('foo.com')
+      expect(tab.url).to.equal('https://foo.com/abc')
+
+      const [callbackOnRemove] = mocks.chrome.tabs.onRemoved.addListener.firstCall.args
+
+      const removedTabId = 18
+      const removeInfo = {
+        id: 18,
+        windowId: 1,
+        active: false,
+        url: 'https://new.com/abc',
+      }
+
+      callbackOnRemove(removedTabId, removeInfo)
+
+      const stateOnRemoved = mocks.getState()
+      const tabsNotRemoved = stateOnRemoved.tabs
+
+      expect(tabsNotRemoved.get(removedTabId)).to.equal(undefined)
+    })
+  })
+
+  describe('chrome.windows.onCreated', () => {
+    it("Adds a newly created window to store's state", () => {
+      const [callback] = mocks.chrome.windows.onCreated.addListener.firstCall.args
+
+      const win = { id: 4, focused: true }
+      callback(win)
+      const state = mocks.getState()
+
+      if (state.mostRecentAction.type !== 'chrome.windows.onCreated') {
```

Comment:
There are 2 cases we want to test. One where `focused: true` and one where `focused: false`. In the first case the focusedWindowId should now be updated, but not in the second case. We assume that the `mostRecentAction` works, so we don't need to test that.

Reply:
Refactor as instructed